iothub/device/tests/ClientPropertyCollectionTestsNewtonsoft.cs

```diff
@@ -0,0 +1,114 @@
+        private const int intPropertyValue = 12345678;
+        private const short shortPropertyValue = 1234;
+        private const string stringPropertyValue = "propertyValue";
+        private const string componentName = "testableComponent";
+        private const string writablePropertyDescription = "testableWritablePropertyDescription";
+        private const string updatedPropertyValue = "updatedPropertyValue";
+        private const string twinJson = "{ \"stringPropertyName\" : \"propertyValue\", \"boolProperty\" : false, \"doublePropertyName\" : 1.001, \"floatProperty\" : 1.2, \"intProperty\" : 12345678, \"shortProperty\" : 1234 }";
+        private const string twinJsonWritableProperty = "{ \"stringPropertyName\" : { \"value\" : \"propertyValue\",  \"ac\" : 200, \"av\" : 2, \"ad\" : \"testableWritablePropertyDescription\" } }";
+        private const string twinJsonWithComponent = "{ \"testableComponent\" : { \"__t\" : \"c\",  \"stringPropertyName\" : \"propertyValue\", \"boolProperty\" : false, \"doublePropertyName\" : 1.001, \"floatProperty\" : 1.2, \"intProperty\" : 12345678, \"shortProperty\" : 1234 } }";
+        private const string twinJsonWritablePropertyWithComponent = "{ \"testableComponent\" : { \"__t\" : \"c\", \"stringPropertyName\" : { \"value\" : \"propertyValue\",  \"ac\" : 200, \"av\" : 2, \"ad\" : \"testableWritablePropertyDescription\" } } }";
+
+        private static TwinCollection collectionToRoundTrip = new TwinCollection(twinJson);
+        private static TwinCollection collectionWritablePropertyToRoundTrip = new TwinCollection(twinJsonWritableProperty);
+        private static TwinCollection collectionWithComponentToRoundTrip = new TwinCollection(twinJsonWithComponent);
+        private static TwinCollection collectionWritablePropertyWithComponentToRoundTrip = new TwinCollection(twinJsonWritablePropertyWithComponent);
+
+        [TestMethod]
+        public void ClientPropertyCollectionNewtonsoft_CanGetSimpleValue()
+        {
+            
+            var clientProperties = ClientPropertyCollection.FromTwinCollection(collectionToRoundTrip, DefaultPayloadConvention.Instance);
+            clientProperties.Convention = DefaultPayloadConvention.Instance;
+            
+            clientProperties.TryGetValue(stringPropertyName, out string stringOutValue);
+            Assert.AreEqual(stringPropertyValue, stringOutValue, $"Property values do not match, expected {stringPropertyValue} but got {stringOutValue}");
+            clientProperties.TryGetValue(boolPropertyName, out bool boolOutValue);
+            Assert.AreEqual(boolPropertyValue, boolOutValue, $"Property values do not match, expected {boolPropertyValue} but got {boolOutValue}");
+            clientProperties.TryGetValue(doublePropertyName, out double doubleOutValue);
+            Assert.AreEqual(doublePropertyValue, doubleOutValue, $"Property values do not match, expected {doublePropertyValue} but got {doubleOutValue}");
```

Comment:
@abhipsa, this fails when trying to convert some of the primitive types.

What happens is the float value type is deserialized from Newtonsoft as a double so the casting doesn't work as expected.

This also happens with int as the base type is Int64.

We will need to force the cast for the specified type or document the default behavior of non typed deserializaion.

Reply:
I added the casts for some specific numeric primitives (float, int32) and the left the default ones (long, string, double) as is. This will allow us to generically handle these. DTDL will dictate that values have a certain format, such as date time, but not for others like long. It's possible we should choose to create very specific methods for **ALL** of the primitives.

| Digital twin primitive schema | Description |
| --- | --- |
|~`boolean`~| ~A boolean value~|
| `date` | A full-date as defined in [section 5.6 of RFC 3339](https://tools.ietf.org/html/rfc3339#section-5.6) |
| `dateTime` | A date-time as defined in [RFC 3339](https://tools.ietf.org/html/rfc3339) |
| ~`double`~ | ~An IEEE 8-byte floating point~ |
| `duration` | A duration in ISO 8601 format |
| ~`float`~ | ~An IEEE 4-byte floating point~ |
| ~`integer`~ | ~A signed 4-byte integer~ |
| ~`long`~ | ~A signed 8-byte integer~ |
| ~`string`~ | ~A UTF8 string~ |
| `time` | A full-time as defined in [section 5.6 of RFC 3339](https://tools.ietf.org/html/rfc3339#section-5.6) |


### NOTE

Consider that a customer could create their own JSON payload with decimal values stored like this `{ "myProp" : "3,14579" }`. When Newtonsoft gets the value it will treat it as a string because of the quotes. If the customer wanted to treat this as a `long` we'd eventually call `Convert.ToLong()` and it would fail with a NumberFormat failure. Or, possibly someone tries to convert `"awesomeString` to a `float`.

In both of these situations we don't handle this conversion so we'll throw an exception from the `Convert` class which is acceptable. But, we need to explore these and document the exceptions and the reasons why.